### ISSUE TITLE
Yolo-V3 Tf-Lite model running with nntreamer pipeline

### ISSUE THREAD

**Issue by amrithkrish** [nnstreamer/nnstreamer#4246]
Hi, 

I'm trying to run a custom Yolo-V3 Tf-Lite model on NXP imx8M plus board using nnstreamer pipeline.

The pipeline I have used is as follows :

`gst-launch-1.0` v4l2src  device=/dev/video3 ! videoconvert ! videoscale ! video/x-raw,width=1000,height=1000,format=RGB,pixel-aspect-ratio=1/1,framerate=30/1 ! tee name=t t. ! queue leaky=2 max-size-buffers=2 ! videoscale ! video/x-raw,width=320,height=320,format=RGB ! tensor_converter ! tensor_transform mode=arithmetic option=typecast:float32,div:255.0 ! queue ! tensor_filter framework=tensorflow2-lite model=/home/model.tflite ! other/tensors,num_tensors=1,types=float32,dimensions=640:480:1,format=static ! tensor_decoder mode=bounding_boxes option1=yolov3 option2=/home/labels.txt option3=0 option4=1000:1000 option5=320:320 ! video/x-raw,width=1000,height=1000,format=RGBA ! mix.sink_0 t. ! queue leaky=2 max-size-buffers=10 ! mix.sink_1 compositor name=mix sink_0::zorder=2 sink_1::zorder=1 ! waylandsink `

However the above pipeline execution is resulting me in the error : 

`WARNING: erroneous pipeline : could not link tensorfilter0 to tensordec0, tensorfilter0 can't handle caps other/tensors, num_tensors=(int)1, types=(string)float32, dimensions=(string)640:480:1, format=(string)static `

 I'm guessing my pipeline and the model have some parameter mismatch. But I am not sure what are the parameters I need to change in the pipeline so as to make it work with YoloV3.

Could someone help me with this?

**Comment by anyj0527** [comment]
The log says the output of `/home/model.tflite` does not match with the `other/tensors,other/tensors,num_tensors=1,types=float32,dimensions=640:480:1,format=static`. You may check your model.
Anyway there is no tensor_decoder for yolov3. You may try popular yolov5 or yolov8 instead: [REF](https://github.com/anyj0527/nnstreamer-example/tree/add-yolov8-bash-example/bash_script/example_yolo)

**Comment by amrithkrish** [comment]
I have to use yolov3 model and not the yolov5 or yolov8.

Apologies, if I understood this wrong, `Anyway there is no tensor_decoder for yolov3` - Does this mean I might not be able to make the YoloV3 model running using nnstreamer?

Or does it mean the pipeline is possible, but the tensor_decoder component should be changed o removed?

Also, what parameters should I change in the pipeline as per my model so as to make it work?

**Comment by myungjoo** [comment]
> Anyway there is no tensor_decoder for yolov3 - Does this mean I might not be able to make the YoloV3 model running using nnstreamer?

No, you can run yolov3 with nnstreamer.
You just cannot VISUALIZE the output without updating the code for yolov3.

In general, AI services do not need visualization.
You need it when you demonstrate to people, not when you send the inference results to a software module.

> Or does it mean the pipeline is possible, but the tensor_decoder component should be changed o removed?

So.. yes, again, the pipeline is still possible.
If you really need to "visualize" the results of yolov3, you can "add" yolov3 deciding logic to the subplugin of bounding-box or write your own "new" subplugin that translates yolov3 results to a video frame.

**Comment by amrithkrish** [comment]
I had tried with Yolo-v5 model with a custom dataset. I exported it to tflite. The model was working fine with real time data ran from my PC.

 Now in order to make the same model to work in imx8m+ board, I had generated the nnstreamer pipeline for it.
The pipeline I tried is as follows :

`gst-launch-1.0 v4l2src  device=/dev/video3 ! jpegdec ! videoscale ! videoconvert ! video/x-raw,width=810,height=1080,format=RGB,framerate=0/1 ! tee name=t t. ! queue ! videoscale ! video/x-raw,width=416,height=416,format=RGB ! tensor_converter ! other/tensors,num_tensors=1,types=uint8,format=static,dimensions=3:416:416:1 ! queue ! tensor_filter framework=tensorflow2-lite model=/home/int8_yolov5.tflite  ! other/tensors,format=static,num_tensors=1,dimensions=85:10647:1:1,types=uint8 ! tensor_transform mode=arithmetic option=typecast:float32,add:-3.0,mul:0.004723627585 ! queue ! tensor_decoder mode=bounding_boxes option1=yolov5 option2=/home/labels.txt option4=810:1080 option5=416:416 ! video/x-raw,width=810,height=1080,format=RGBA ! mix.sink_0  t. ! queue ! mix.sink_1 compositor name=mix sink_0::zorder=2 sink_1::zorder=1 ! videoconvert ! waylandsink`


The above pipeline is however giving me the error as follows -
`
(gst-launch-1.0:1767): CRITICAL **: bb_getOutCaps : assertion 'dim[0] == (data->labeldata.total_labels + YOLOV5_DETECTION_NUM_INFO)' failed  
WARNING : erroneous pipeline: Could not link queue1 to tensorfilter0`


Can someone please tell me what should I change in my pipeline so that it will work from the imx8m+ board?

**Comment by amrithkrish** [comment]
Hi,

I tried following the link - [https://github.com/anyj0527/nnstreamer-example/blob/add-yolov8-bash-example/bash_script/example_yolo/gst-launch-object-detection-yolov5-tflite.sh] for using Yolo-v5 fp-16 Tf-Lite model for object detection. 
Sorry, I am new to this topic. Can someone tell me how did the dimension parameter - `dimensions=85:6300:1` was obtained?

Thanks in advance.

**Comment by Nicolas-Gsln** [comment]
Hi,

About the dimension parameter required :
- 85 is the number of parameters that the model should output (there are 80 values corresponding to the confidence for each of the 80 classes, and 5 parameters for x and y coordinates, height and width of the box and the confidence that this cell contains an object).
dim[0] must be : data->labeldata.total_labels(80) + YOLOV5_DETECTION_NUM_INFO(5) = 80+5 = 85
- The 6300 parameters are linked to the dimension of your input image and the strides of Yolov5 (32, 16 and 8) :
dim[1] must be max_detection, which is :
    data->max_detection = (
        (data->i_width / 32) * (data->i_height / 32) +
        (data->i_width / 16) * (data->i_height / 16) +
        (data->i_width / 8) * (data->i_height / 8)) * 3;
In your case, ((320/32)²+(320/16)²+(320/8)²)*3 = 6300

So it seems that the output of your model is not corresponding to the expected input of the tensor_decoder, maybe you don't apply a part of the post-processing to your model according to the tensor_decoder expect format ?

We were able to run a Yolov4-tiny model using the Yolov5 tensor_decoder with a python filter between the tensor_filter and the tensor_decoder to apply the missing post-processing (in our case the sigmoid function was not applied and a dimension was missing).
Maybe it could help you to see [how we did the post-processing](https://github.com/nxp-imx/nxp-nnstreamer-examples/blob/main/detection/postprocess_yolov4_tiny.py) ?

Regards.
